arch/x86/kvm/pkvm/vmx/irq.c

```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: GPL-2.0
+#include <asm/ptrace.h>
+#include <asm/trapnr.h>
+#include <vmx/x86_ops.h>
+#include <pkvm.h>
+#include "pkvm/irq.h"
+
+static void handle_nmi(struct pt_regs *regs, int vector, bool has_error_code)
+{
+	int cpu = raw_smp_processor_id();
+	struct pkvm_host_vcpu *hvcpu;
+
+	hvcpu = pkvm_hyp->host_vm.host_vcpus[cpu];
+	if (!hvcpu || hvcpu->pending_nmi)
+		return;
+	/*
+	 * This NMI could happen either before executing
+	 * the injection code or after.
+	 * For the before case, should record a pending NMI.
+	 * For the after case, if no NMI is injected in guest
+	 * we also need to record a pending NMI. If NMI is
+	 * injected already, it is not necessary to inject
+	 * again but injecting it in the next round should also
+	 * be fine. So simply record a pending NMI here.
```

Comment:
This comment it not quite clear to me. What exactly is "injection code"? The code in handle_pending_events() which handles a previous NMI? And what exactly is "before" and "after"? Before and after which exact moment in time?

Reply:
Sounds good to me (except that I'm not sure what exactly "in this run loop" means).

The 2nd sentence probably belongs not here but below, before vmx_enable_irq_window()? (maybe merge it with the existing comment there, to make it all sound a bit clearer)

Reply:
"in this run loop" - I would like to mean by in this host vmexit handling cycle. Probably I can just say this which may be more clear?

And yes, will merge the comments.